### PR TITLE
React router

### PR DESCRIPTION
**Pull request by chrisvanmook**

### CODE REVIEW

File: src/params.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+export interface UrlParam<T> {
```
Comment (jonkoops) [inline src/params.ts]: I think we can place the type definition in `useParam.ts` since it's used there.

File: yarn.lock [hunk 2]
```diff
@@ -1004,7 +1004,7 @@
   dependencies:
```
Comment (jonkoops) [inline yarn.lock:1004]: We already have a `package-lock.json` file

File: src/useParam.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        : encodedValue
+
+    if (newValue) {
+      params.set(urlParam.name, newValue)
+    } else {
+      params.delete(urlParam.name)
+    }
+
+    // We don't want the order to change, so always sort them before updating the URL
+    params.sort()
```
Comment (jonkoops) [inline src/useParam.ts:36]: That's pretty cool, didn't know about this little method.

File: src/useParam.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  const location = useLocation()
+  const params = new URLSearchParams(location.search)
+  const rawValue = params.get(urlParam.name)
+
+  const setValue: SetValueFn<T> = (value, method = 'push') => {
+    const encodedValue = urlParam.encode(value)
+
+    // Check if we don't want to show the default value in the URL
+    const newValue =
+      !urlParam.showDefaultValueInQuery && deepEqual(urlParam.defaultValue, value)
```
Comment (jonkoops) [inline src/useParam.ts:25]: In what scenario would we want to use `showDefaultValueInQuery`? I would leave it our for now until we actually run into a scenario where we would need it.

File: src/useParam.test.tsx [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  name: 'complex',
+  defaultValue: [
+    [1, 1],
+    [2, 2],
+  ],
+  encode: (value) => JSON.stringify(value),
+  decode: (value) => JSON.parse(value),
+}
+
+const Component = () => {
```
Comment (jonkoops) [inline src/useParam.test.tsx]: Do we need a component to make this work? Could we not use something like [@testing-library/react-hooks](https://github.com/testing-library/react-hooks-testing-library), like we are in Atlas?

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: src/useParam.ts (hunk 4 above)
Comment (chrisvanmook) [reply src/useParam.ts:25]: Indeed we don't need it for now, but when we publish this as a package, i'd be sure this is a needed feature. So I'll leave it in for now

File: yarn.lock (hunk 2 above)
Comment (chrisvanmook) [reply yarn.lock:1004]: removed it

**Review by jonkoops — APPROVED** [review]

**Pull request merged**